### ISSUE TITLE
Map ts.SyntaxKind to corresponding wrapped node

### ISSUE THREAD

**Issue by sfrooster** [dsherret/ts-simple-ast#36]
Trying to use this lib in a TypeScript project of my own and having difficulty taking advantage of the typings.  For example, the following code works (it's just something i'm hacking at right now):

```typescript
import Ast from "ts-simple-ast";
// import { NamespaceDeclaration } from "ts-simple-ast";
import * as ts from "typescript";
// import * as _ from "lodash";

const ast = new Ast({
    tsConfigFilePath: "tsconfig.json"
});

ast.addSourceFiles("msgs/**/*{.d.ts,.ts}");

ast.getSourceFiles().forEach((sf, idx) => {
    if (idx === 0) {
        console.log(sf.getFilePath());
        sf.getNamespaces().forEach(ns => processNS([], ns));
        console.log("=================");
    }
});

function processNS(nsParents, ns) {
    let currNSStr = nsParents.length === 0 ? ns.node.name.text : nsParents.map(ns => ns.node.name.text).join(".");
    ns.getClasses().forEach(c => {
        console.log(`${currNSStr}.${c.node.name.text}`);
    });

    let currNS = nsParents.concat(ns);
    ns.getChildrenOfKind(ts.SyntaxKind.ModuleDeclaration).forEach(childNS => {
        processNS(currNS, childNS);
    });
}
```
The problem is I'd like, for example, for the function to be written as follows:
```typescript
function processNS(nsParents, ns: NamespaceDeclaration) {
    let currNSStr = nsParents.length === 0 ? ns.node.name.text : nsParents.map(ns => ns.node.name.text).join(".");
    ns.getClasses().forEach(c => {
        console.log(`${currNSStr}.${c.node.name.text}`);
    });

    let currNS = nsParents.concat(ns);
    ns.getChildrenOfKind(ts.SyntaxKind.ModuleDeclaration).forEach(childNS => {
        processNS(currNS, childNS);
    });
}
```
But then vscode complains about the "node" portion of "ns.node.name.text" and "c.node.name.text" (and it also complains about "childNS" which it says is of type Node<ts.Node>).

If I change it instead to:
```typescript
function processNS(nsParents, ns: ts.NamespaceDeclaration) {
    let currNSStr = nsParents.length === 0 ? ns.node.name.text : nsParents.map(ns => ns.node.name.text).join(".");
    ns.getClasses().forEach(c => {
        console.log(`${currNSStr}.${c.node.name.text}`);
    });

    let currNS = nsParents.concat(ns);
    ns.getChildrenOfKind(ts.SyntaxKind.ModuleDeclaration).forEach(childNS => {
        processNS(currNS, childNS);
    });
}
```
Then it also complains about the "node" in "ns.node.name.text" but also "getClasses" of "ns.getClasses()" and "getChildrenOfKind" of "ns.getChildrenOfKind().

So, the question is - how should I be setting that type such that I can take advantage of the typing system?

**Comment by dsherret** [comment]
Hey Michael, in 0.29.0 I just did the following change:

1. Removed `getCompilerNode()` method.
2. Changed internal `.node` property to `.compilerNode`.

So if you access `.compilerNode` instead of `.node` the above code should now work.

----

Regarding the return type of `.getChildrenOfKind()` and such methods... I've been thinking about that as well. Unfortunately, I don't currently have a mapping for syntax kinds to class types. I'm not sure an easy way to do that, but what I was thinking is to add method signatures for all the possibilities, like so:

```typescript
getChildrenOfKind(kind: ts.SyntaxKind.ModuleDeclaration): NamespaceDeclaration[];
getChildrenOfKind(kind: ts.SyntaxKind.ClassDeclaration): ClassDeclaration[];
```

That would be code generated so it's maintainable. I'd just have to maintain a list of all the mappings. Then that code generation should also be applied in other areas like `getFirstChildOfKind`.

For now, you can use the type argument in order to save yourself a type assertion, but it's not ideal:

```typescript
ns.getChildrenOfKind<NamespaceDeclaration>(ts.SyntaxKind.ModuleDeclaration).forEach(childNs => {
    // childNs will be NamespaceDeclaration
});
```

I'm going to mark this issue as low priority and get to it later on.

**Comment by dsherret** [comment]
This will be implemented in 0.59.0.

It's kind of cool... the specific overloads are code generated in the definition file using this library. The Node.d.ts file size goes from about 250 lines to over 3000 with all the specific overloads and js docs.

----------------

I also added a `.getNodeProperty` method that can be used to get a node property off the underlying compiler node as a wrapped node (couldn't think of a better name).

For example:

```typescript
ns.getNodeProperty("name"); //returns Node<ts.Identifier>
```
